File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = []
    def pre(self, root: Optional[TreeNode]) -> None:
        if root == None:
            return
        self.ans.append(root.val)
        self.pre(root.left)
        self.pre(root.right)
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.pre(root)
        return self.ans
        

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if(root == NULL) return {};
        vector<int> ans;
        stack<TreeNode*> st;
        st.push(root);

        while(!st.empty())
        {
            auto top = st.top();
            st.pop();
            ans.push_back(top->val);
            if(top->right) st.push(top->right);
            if(top->left) st.push(top->left);
            
        }
        return ans;

    }
};